Decomposer extraction for Rust native `ConsolidateBlocks`

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
The following commit adds the ability for the `ConsolidateBlocks` pass to instantiate decomposers and choose a basis gate based on the provided `Target`. In the case that a `Target` is not provided, it will default to a `TwoBasisDecomposer` with a `CX` gate.


### Details and comments
Built on top of #14748. Will rebase after it merges.
Prerequisite for #14751.

raynelfss committed:
Add: Decomposer extraction for Rust native `ConsolidateBlocks`.
Built on top of #14748.
The following commit adds the ability for the `ConsolidateBlocks` pass to instantiate decomposers and choose a basis gate based on the provided `Target`. In  the case that a `Target` is not provided, it will default to a `TwoBasisDecomposer` with a `CX` gate.

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -44,6 +50,103 @@ pub enum DecomposerType {
+
+/// The collection of compatible KAK decomposition parametric gates in a set
+/// in order of priority.
+static KAK_GATES_PARAM: OnceLock<IndexSet<StandardGate>> = OnceLock::new();
```
Why do you need a set here? One is 8 items the other is 4 just use a static array, doing `contains()` is a u8 comparison doing that worst case 8 times isn't going to have any real speed difference for this function. The constant factor overhead for building the set and hashing will likely be higher if anything even if the lookup is O(1) vs O(n).

raynelfss replied:
Implemented in 00668e9b711e780d30ea123d703ab06cd5a47de1

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -44,6 +50,103 @@ pub enum DecomposerType {
+        let target_basis_gates: IndexSet<StandardGate> = target
+            .operations()
+            .filter_map(|op| op.operation.try_standard_gate())
+            .collect();
```
Why do you need a collection here? I feel like the allocation overhead here isn't worth it since this should be trivial to check if the gate is supported in the target which is already allocated.

raynelfss replied:
Implemented in 00668e9b711e780d30ea123d703ab06cd5a47de1

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -44,6 +50,103 @@ pub enum DecomposerType {
+                SmallVec::default(),
+                gate.matrix(&[])
+                    .expect("Error while obtaining the matrix form of gate 'cx' without params.")
+                    .view(),
```
Why not just pull directly from the static cx matrix and use aview from the static slice?

raynelfss replied:
Implemented in 00668e9b711e780d30ea123d703ab06cd5a47de1

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -352,7 +455,38 @@ pub fn run_consolidate_blocks(
+/// This is the main function of the `ConsolidateBlocks` transpiler pass
+/// which replaces uninterrupted sequences of gates acting on the same qubits
+/// into a Unitary node, which will consecutively be resynthesized into a
+/// more optimal subcircuit.
```
```suggestion
/// This is function is the Rust entry point for the `ConsolidateBlocks` transpiler pass
/// which replaces uninterrupted sequences of gates acting on the same pair of qubits
/// into a [`UnitaryGate`] representing the unitary of that two qubit block if it estimated to
/// to optimize the circuit. This [`UnitaryGate`] subsequently will be synthesized by the
/// unitary synthesis pass into a more optimal subcircuit to replace that block.
```

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -352,7 +455,38 @@ pub fn run_consolidate_blocks(
+/// * `target` - The target representing the backend for which the pass is consolidating.
+pub fn run_consolidate_blocks(
+    dag: &mut DAGCircuit,
+    approximation_degree: f64,
```
Isn't `approximation_degree` normally an `Option<f64>` where `None` means use the error rate specified in the `Target` for the basis gate on those qubits?

raynelfss replied:
In `ConsolidateBlocks` it just defaults to 1.0 and is only used for the basis fidelity of TwoQubitBasisDecomposer. I can make it optional so that it regularly defaults to 1.0 if anything.

mtreinish replied:
Hmm, yeah it looks like it's all `f64` in that code path. We definitely have code that works with an option but if it's not supported in the pass's code then best not introduce it in this PR.

raynelfss committed:
Chore: Avoid unnecessary collections and remove OnceCell.
- Use CX matrix directly with aview2

raynelfss committed:
Update crates/transpiler/src/passes/consolidate_blocks.rs

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

raynelfss committed:
Fix: Make `approximation_degree` optional.

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -44,6 +47,91 @@ pub enum DecomposerType {
+        // Targets from C should only support Standard gates.
+        if let Some(gate) = KAK_GATES_PARAM
+            .iter()
+            .find(|gate| target.contains_key(gate.name()))
```
Was there a reason you chose to use a string match here? You should be able to match on the operations from the target right? As a rule of thumb any time you can match on the `StandardGate` directly that's the better path. Like I think the most efficient check would be to iterate over the operations in the target and run `matches!(gate, StandardGate::RXX | StandardGate::RZZ, ...)`.

Also this isn't validating that that gate is actually the correct gate. If I did something like: `target.add_instruction(RZZGate(3.14))` this would match this but there isn't actually an appropriate gate in the target because `ControlledUDecomposer` only supports parameterized gates.

raynelfss replied:
Well, the comparison in the python side `ConsolidateBlocks` is mainly done by gate name. I typically do prefer using the `StandardGate` comparison method. So, if there won't be any sacrifica made by doing a lookup vs performing an O(n) comparison (where n is the number of operations in the `Target`), then that should be okay.

As for the parameter comparison. The Python side `ConsolidateBlocks` also does something curious, since it obtains its `basis_gates` as strings, it doesn't take into consideration parameters at all. In fact the mapping `KAK_GATES_PARAM_NAMES` contains a mapping between the gate names and their class alias, as opposed to `KAK_GATES_NAMES` which maps between names and instances of gates. I don't see how it decides between a gate with a fixed parameter and a parametrized gates without it involving some alias within the `Target`. If this is not the expected behavior, adding an extra check in Rust should be easy enough to know whether the gate has fixed parameters or not.

mtreinish replied:
Honestly I'm not sure it matters what Python does, we're building a rust/C only interface. I expect Python is being looser here because realistically this is just used as an estimate on whether synthesis will result in fewer 2q gates even if an unsupported gate is supported it doesn't produce invalid results, it just means that either the output isn't as optimized as possible (either because we're re-synthesizing the block when it's less efficient, or because we're not synthesizing when it would be beneficial). #13419 will simplify this, but it's not ready yet and also we still will be using `ConsolidateBlocks` in the init stage, although there it runs without a target.

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -44,6 +47,91 @@ pub enum DecomposerType {
+                TwoQubitBasisDecomposer::new_inner(
+                    gate.into(),
+                    SmallVec::default(),
+                    gate.matrix(&[]).unwrap_or_else(|| panic!("Error while obtaining the matrix form of gate '{}' without params.", gate.name())).view(),
```
All of the gates used here have static matrices, I wonder if you could use them all directly in like a big match or something. Not sure how that will look in practice it might be uglier than I think.

raynelfss replied:
At that point, I don't think it is worth doing that. The call here is for gates that are non-parametric, so it still ends up calling the same methods. Do you think there are performance implications with the extra calls? If so, maybe we just need to inline the `Operation::matrix()` method instead.

mtreinish replied:
`gate.matrix()` allocates a new matrix because it returns an owned `Array2`. If you used the static slice to build an aview2 you are avoiding an allocation.

raynelfss replied:
Fixed in c1b3619b1f853d6148539a27991b700bc0f12b81

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -352,7 +440,40 @@ pub fn run_consolidate_blocks(
+    approximation_degree: Option<f64>,
+    target: Option<&Target>,
+) -> PyResult<()> {
+    let approximation_degree = approximation_degree.unwrap_or(1.0);
```
If you're making this an optional, `None` for `approximation_degree` normally is used to indicate error up to the target's error rate. That is how the parameter is used for all the pass which do approximation (which consolidate blocks should be doing so it's estimate is more accurate, although as you commented it doesn't currently).

raynelfss committed:
Fix: Use `StandardGate` comparison instead of string comparisons.

raynelfss committed:
Fix: Use matrix views

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -44,6 +50,107 @@ pub enum DecomposerType {
+    StandardGate::CRX,
+    StandardGate::CRY,
+    StandardGate::CRZ,
+];
```
You can just use: https://github.com/Qiskit/qiskit/blob/main/crates/transpiler/src/passes/unitary_synthesis.rs#L74 and https://github.com/Qiskit/qiskit/blob/main/crates/transpiler/src/passes/unitary_synthesis.rs#L101 instead of redefining these here.

raynelfss replied:
Added in c41c85036a932f21e2e394df7f9720a2294f700b

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -44,6 +50,107 @@ pub enum DecomposerType {
+        StandardGate::CX => aview2(&CX_GATE),
+        StandardGate::CZ => aview2(&CZ_GATE),
+        StandardGate::ECR => aview2(&ECR_GATE),
+        StandardGate::ISwap => aview2(&ISWAP_GATE),
```
With the above suggestion you will need to expand this match to cover the extra gates

raynelfss replied:
Fixed in c41c85036a932f21e2e394df7f9720a2294f700b

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -44,6 +50,107 @@ pub enum DecomposerType {
+                            "Error while creating Basis Decomposer using a {} gate.",
+                            gate.name()
+                        )
+                    }),
```
Why not just use `.expect()`?

raynelfss replied:
I think I was originally using `.expect()` but `clippy` changes it to this to avoid allocating a `String` during the `format!()` call.

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -352,7 +459,40 @@ pub fn run_consolidate_blocks(
+/// * `force_consolidate` - Decides whether to force all consolidations or not.
+/// * `approximation_degree` - A float between `[0.0, 1.0]`. Lower approximates more.
+/// * `target` - The target representing the backend for which the pass is consolidating.
+pub fn run_consolidate_blocks(
```
Can we have a rust test that exercises this function. I realize there is a follow on PR testing it from C, but having a rust test here is good to make sure we have some confidence this rust interface works as expected. Especially since there will likely be a time gap between the two PRs and we don't want this to potentially break.

raynelfss replied:
Sure!

raynelfss replied:
Added a test in ac5baa1bcb095cf11370cdc3807c7707f350fc0a

raynelfss committed:
Fix: Use defined basis gates from `unitary_synthesis`.
- Use already defined `PARAM_SET_BASIS_GATES` and `TWO_QUBIT_BASIS_SET_GATES` and modify the `get_matrix` method to include those as well.
- Make `PARAM_SET_BASIS_GATES` and `TWO_QUBIT_BASIS_SET_GATES` pub.

raynelfss committed:
Test: Add test case for rust entry point.

raynelfss committed:
Fix: Add no miri flag to tests

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -44,6 +52,88 @@ pub enum DecomposerType {
+        if let Some(gate) = target.operations().find_map(|op| {
+            op.operation
+                .try_standard_gate()
+                .and_then(|gate| PARAM_SET_BASIS_GATES.contains(&gate).then_some(gate))
```
If you actually did:
```suggestion
                .and_then(|gate| matches!(gate, PARAM_SET!())).then_some(gate))
```
and pulling the macro in form unitary synthesis. I think it would be potentially slightly more efficient, but it's really a small micro optimization.

raynelfss replied:
Fixed in cb637cc3f74f3788ad8b8108c0f13bc222c87c35

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -44,6 +52,88 @@ pub enum DecomposerType {
+        if let Some(gate) = target.operations().find_map(|op| {
+            op.operation
+                .try_standard_gate()
+                .and_then(|gate| TWO_QUBIT_BASIS_SET_GATES.contains(&gate).then_some(gate))
```
Same comment here but with `TWO_QUBIT_BASIS_SET` instead.

```suggestion
                .and_then(|gate| matches!(gate TWO_QUBIT_BASIS_SET!())).then_some(gate))
```

raynelfss replied:
Fixed in cb637cc3f74f3788ad8b8108c0f13bc222c87c35

mtreinish reviewed (commented):
One small comment on a micro optimization we can do, but otherwise I think this is about ready.

raynelfss committed:
Fix: Apply code suggestions

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

mtreinish commented on crates/transpiler/src/passes/unitary_synthesis.rs:
```diff
@@ -56,6 +56,7 @@ const PI4: f64 = PI / 4.;
 
 /// The matcher for the set of standard gates that the TwoQubitControlledUDecomposer
 /// supports
+#[macro_export]
```
I looked into the details here you can just export the macro within the crate by adding `pub(crate) use PARAM_SET` to within this module file. That will re-export the macro with crate visibility. I don't think we should use `macro_export` here because it means the macro is exposed at the crate boundary which seems too public. See: https://doc.rust-lang.org/edition-guide/rust-2018/path-changes.html#macros

raynelfss replied:
Applied in e3b8e85583369eb9e3a84a237a21a9e4aeb9e750

raynelfss committed:
Fix: Correct import of macros from `unitary_synthesis`.

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -44,6 +52,88 @@ pub enum DecomposerType {
+                    TwoQubitControlledUDecomposer::new(RXXEquivalent::Standard(gate), "ZXZ")
+                        .unwrap_or_else(|_| {
+                            panic!(
+                                "Error while creating Controlled U decomposer using a {} gate.",
```
```suggestion
                                "Error while creating TwoQubitControlledUDecomposer using a {} gate.",
```

raynelfss replied:
Addressed in e08cd6b91ad5614bddf77ec94ccf26fef8daddd5

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -44,6 +52,88 @@ pub enum DecomposerType {
+                    )
+                    .unwrap_or_else(|_| {
+                        panic!(
+                            "Error while creating Basis Decomposer using a {} gate.",
```
```suggestion
                            "Error while creating TwoQubitBasisDecomposer using a {} gate.",
```

raynelfss replied:
Addressed in e08cd6b91ad5614bddf77ec94ccf26fef8daddd5

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -44,6 +52,88 @@ pub enum DecomposerType {
+                "U",
+                None,
+            )
+            .expect("Error while creating Basis Decomposer using a 'cx' gate."),
```
```suggestion
            .expect("Error while creating TwoQubitBasisDecomposer using a 'cx' gate."),
```

raynelfss replied:
Addressed in e08cd6b91ad5614bddf77ec94ccf26fef8daddd5

mtreinish commented on crates/transpiler/src/passes/consolidate_blocks.rs:
```diff
@@ -352,7 +442,189 @@ pub fn run_consolidate_blocks(
+            }
+        } else {
+            panic!("The resulting circuit was unexpectedly empty.");
+        };
```
I think for clarity and more thorough checking I'd write the test like:

```suggestion
        let data = circuit_result.data();
        if data.len() != 1 {
            panic!("The output circuit had {data.len()} instructions but there should only be 1 unitary gate");
        }
        let OperationRef::Unitary(unitary) = data.first().unwrap().op.view() else {
            panic!("The output circuit didn't result in a single unitary gate");
        }
        let expected = aview2([
            [...],
            [...],
        ])
        assert!(abs_diff_eq(unitary.matrix_view(), expected));
```
Check my syntax here since the github webui is where syntax errors are injected and also filling in the matrix with the expected value.

raynelfss replied:
I ended up simplifying the test to just an identity cancellations as we discussed before.

mtreinish reviewed (commented):
Just some small last comments, then I'm pretty sure I'll approve it after that.

raynelfss committed:
Chore: Simplify test

raynelfss committed:
FIx: Miri flag
